core/src/main/java/org/infinispan/marshall/core/ClassIdentifiers.java

```diff
@@ -0,0 +1,71 @@
+package org.infinispan.marshall.core;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.infinispan.configuration.global.GlobalConfiguration;
+import org.infinispan.container.entries.InternalCacheValue;
+
+public class ClassIdentifiers implements InternalClassIds {
+   private final static ClassIdentifiers INSTANCE = new ClassIdentifiers();
+
+   // the hashmap is not changed after static ctor, therefore concurrent access is safe
+   private final Map<Class<?>, Integer> classToId = new HashMap<>();
```

Comment:
> Though, with all due respect, the implementation in ClassToExternalizerMap sucks, too. Solving collisions by just going on the next spot in the table generates a sequences of often colliding elements. You even don't count how many collisions there are (to resize when you reach a threshold).

To most important use of `ClassToExternalizerMap` is for internal externalizers, which we know exactly how many they are. I've sized that collection with the right size and load factor to avoid collisions as much as possible, so although you highlight a problem, in practical terms I don't expect it to happen.

> Seems I'll have to implement a perfect hashing table to comfort you here...

I didn't ask for perfect hashing table :). I'm just asking that given the elements added to this map and its use of, we can have something a bit more lightweight than a `HashMap`.